In [1]:
import numpy as np
import tensorflow as tf

keras = tf.keras

# Gradient Vanishing problem & glorot and HE INITILIALIZATION

In [2]:
keras.layers.Dense(10, activation="relu", kernel_initializer="he_normal")

In [3]:
he_avg_unit = keras.initializers.VarianceScaling(scale=2., mode='fan_avg', distribution='uniform')
keras.layers.Dense(10, activation="sigmoid", kernel_initializer=he_avg_unit)

In [4]:
# glorot and he initialization and all initializators ; this is all 
[name for name in dir(keras.initializers) if not name.startswith("_")]

['Constant',
 'GlorotNormal',
 'GlorotUniform',
 'HeNormal',
 'HeUniform',
 'Identity',
 'Initializer',
 'LecunNormal',
 'LecunUniform',
 'Ones',
 'Orthogonal',
 'RandomNormal',
 'RandomUniform',
 'TruncatedNormal',
 'VarianceScaling',
 'Zeros',
 'constant',
 'deserialize',
 'get',
 'glorot_normal',
 'glorot_uniform',
 'he_normal',
 'he_uniform',
 'identity',
 'lecun_normal',
 'lecun_uniform',
 'ones',
 'orthogonal',
 'random_normal',
 'random_uniform',
 'serialize',
 'truncated_normal',
 'variance_scaling',
 'zeros']

# Nonsaturatin activation functions && use leaky ReLU 

In [5]:
[m for m in dir(keras.activations) if not m.startswith("_")]


['deserialize',
 'elu',
 'exponential',
 'gelu',
 'get',
 'hard_sigmoid',
 'linear',
 'relu',
 'selu',
 'serialize',
 'sigmoid',
 'softmax',
 'softplus',
 'softsign',
 'swish',
 'tanh']

In [6]:
[m for m in dir(keras.layers) if "relu" in m.lower()]
#If you want use SELU activation set activation="selu" and kernel_initializer="lecum_normal" when creating a layer

['LeakyReLU', 'PReLU', 'ReLU', 'ThresholdedReLU']

In [7]:

(X_train_full, y_train_full), (X_test, y_test) = keras.datasets.fashion_mnist.load_data()
X_train_full = X_train_full / 255.0
X_test = X_test / 255.0
X_valid, X_train = X_train_full[:5000], X_train_full[5000:]
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, kernel_initializer="he_normal"),
    keras.layers.LeakyReLU(),
    keras.layers.Dense(100, kernel_initializer="he_normal"),
    keras.layers.LeakyReLU(),
    keras.layers.Dense(10, activation="softmax")
])

Metal device set to: Apple M1

systemMemory: 16.00 GB
maxCacheSize: 5.33 GB



In [8]:

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.legacy.SGD(learning_rate=1e-3),
              metrics=["accuracy"])

In [9]:
history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10


2023-06-15 20:55:50.173477: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


1719/1719 [==============================] - 10s 6ms/step - loss: 1.2227 - accuracy: 0.6331 - val_loss: 0.8334 - val_accuracy: 0.7444
Epoch 2/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.7578 - accuracy: 0.7582 - val_loss: 0.6797 - val_accuracy: 0.7846
Epoch 3/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.6573 - accuracy: 0.7863 - val_loss: 0.6215 - val_accuracy: 0.7998
Epoch 4/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.6043 - accuracy: 0.8023 - val_loss: 0.5719 - val_accuracy: 0.8146
Epoch 5/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.5703 - accuracy: 0.8120 - val_loss: 0.5440 - val_accuracy: 0.8248
Epoch 6/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.5458 - accuracy: 0.8177 - val_loss: 0.5232 - val_accuracy: 0.8320
Epoch 7/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.5271 - accuracy: 0.8225 - val_loss: 0.5059 - val_accuracy: 0.8

In [10]:
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, kernel_initializer="he_normal"),
    keras.layers.PReLU(),
    keras.layers.Dense(100, kernel_initializer="he_normal"),
    keras.layers.PReLU(),
    keras.layers.Dense(10, activation="softmax")
])

In [11]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.legacy.SGD(learning_rate=1e-3),
              metrics=["accuracy"])

In [12]:
history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))


Epoch 1/10
1719/1719 [==============================] - 12s 7ms/step - loss: 1.4223 - accuracy: 0.5703 - val_loss: 0.9585 - val_accuracy: 0.7040
Epoch 2/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.8444 - accuracy: 0.7306 - val_loss: 0.7479 - val_accuracy: 0.7610
Epoch 3/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.7081 - accuracy: 0.7714 - val_loss: 0.6650 - val_accuracy: 0.7864
Epoch 4/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.6391 - accuracy: 0.7920 - val_loss: 0.6064 - val_accuracy: 0.8014
Epoch 5/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.5952 - accuracy: 0.8042 - val_loss: 0.5709 - val_accuracy: 0.8140
Epoch 6/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.5641 - accuracy: 0.8120 - val_loss: 0.5448 - val_accuracy: 0.8222
Epoch 7/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.5410 - accuracy: 0.8182 - val_loss: 0.5246 - val_ac

# ELU


In [13]:
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt

def elu(z, alpha=1):
    return np.where(z < 0, alpha * (np.exp(z) - 1), z)

In [14]:
plt.plot(z, elu(z), "b-", linewidth=2)
plt.plot([-5, 5], [0, 0], 'k-')
plt.plot([-5, 5], [-1, -1], 'k--')
plt.plot([0, 0], [-2.2, 3.2], 'k-')
plt.grid(True)
plt.title(r"ELU activation function ($\alpha=1$)", fontsize=14)
plt.axis([-5, 5, -2.2, 3.2])

plt.savefig("elu_plot")
plt.show()

NameError: name 'z' is not defined

In [ ]:
keras.layers.Dense(10, activation="elu")

# SELU
This activation function was proposed in this great paper by Günter Klambauer, Thomas Unterthiner and Andreas Mayr, published in June 2017. During training, a neural network composed exclusively of a stack of dense layers using the SELU activation function and LeCun initialization will self-normalize: the output of each layer will tend to preserve the same mean and variance during training, which solves the vanishing/exploding gradients problem. As a result, this activation function outperforms the other activation functions very significantly for such neural nets, so you should really try it out. Unfortunately, the self-normalizing property of the SELU activation function is easily broken: you cannot use ℓ1 or ℓ2 regularization, regular dropout, max-norm, skip connections or other non-sequential topologies (so recurrent neural networks won't self-normalize). However, in practice it works quite well with sequential CNNs. If you break self-normalization, SELU will not necessarily outperform other activation functions.

In [ ]:
from scipy.special import erfc

# alpha and scale to self normalize with mean 0 and standard deviation 1
# (see equation 14 in the paper):
alpha_0_1 = -np.sqrt(2 / np.pi) / (erfc(1/np.sqrt(2)) * np.exp(1/2) - 1)
scale_0_1 = (1 - erfc(1 / np.sqrt(2)) * np.sqrt(np.e)) * np.sqrt(2 * np.pi) * (2 * erfc(np.sqrt(2))*np.e**2 + np.pi*erfc(1/np.sqrt(2))**2*np.e - 2*(2+np.pi)*erfc(1/np.sqrt(2))*np.sqrt(np.e)+np.pi+2)**(-1/2)

In [ ]:
def selu(z, scale=scale_0_1, alpha=alpha_0_1):
    return scale * elu(z, alpha)

In [ ]:
np.random.seed(42)
Z = np.random.normal(size=(500, 100)) # standardized inputs
for layer in range(1000):
    W = np.random.normal(size=(100, 100), scale=np.sqrt(1 / 100)) # LeCun initialization
    Z = selu(np.dot(Z, W))
    means = np.mean(Z, axis=0).mean()
    stds = np.std(Z, axis=0).mean()
    if layer % 100 == 0:
        print("Layer {}: mean {:.2f}, std deviation {:.2f}".format(layer, means, stds))


Layer 0: mean -0.00, std deviation 1.00
Layer 100: mean 0.02, std deviation 0.96
Layer 200: mean 0.01, std deviation 0.90
Layer 300: mean -0.02, std deviation 0.92
Layer 400: mean 0.05, std deviation 0.89
Layer 500: mean 0.01, std deviation 0.93
Layer 600: mean 0.02, std deviation 0.92
Layer 700: mean -0.02, std deviation 0.90
Layer 800: mean 0.05, std deviation 0.83
Layer 900: mean 0.02, std deviation 1.00


By default, the SELU hyperparameters (scale and alpha) are tuned in such a way that the mean output of each neuron remains close to 0, and the standard deviation remains close to 1 (assuming the inputs are standardized with mean 0 and standard deviation 1 too). Using this activation function, even a 1,000 layer deep neural network preserves roughly mean 0 and standard deviation 1 across all layers, avoiding the exploding/vanishing gradients problem:

In [ ]:
keras.layers.Dense(10, activation="selu",
                   kernel_initializer="lecun_normal")

In [ ]:

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(300, activation="selu",
                             kernel_initializer="lecun_normal"))
for layer in range(99):
    model.add(keras.layers.Dense(100, activation="selu",
                                 kernel_initializer="lecun_normal"))
model.add(keras.layers.Dense(10, activation="softmax"))

In [ ]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.legacy.SGD(learning_rate=1e-3),
              metrics=["accuracy"])

In [ ]:
pixel_means = X_train.mean(axis=0, keepdims=True)
pixel_stds = X_train.std(axis=0, keepdims=True)
X_train_scaled = (X_train - pixel_means) / pixel_stds
X_valid_scaled = (X_valid - pixel_means) / pixel_stds
X_test_scaled = (X_test - pixel_means) / pixel_stds

In [ ]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(300, activation="relu", kernel_initializer="he_normal"))
for layer in range(99):
    model.add(keras.layers.Dense(100, activation="relu", kernel_initializer="he_normal"))
model.add(keras.layers.Dense(10, activation="softmax"))

In [ ]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.legacy.SGD(learning_rate=1e-3),
              metrics=["accuracy"])

In [ ]:
history = model.fit(X_train_scaled, y_train, epochs=5,
                    validation_data=(X_valid_scaled, y_valid))

Epoch 1/5
1719/1719 [==============================] - 76s 43ms/step - loss: 1.8380 - accuracy: 0.2573 - val_loss: 1.5318 - val_accuracy: 0.3406
Epoch 2/5
1719/1719 [==============================] - 73s 42ms/step - loss: 1.3377 - accuracy: 0.4124 - val_loss: 1.0494 - val_accuracy: 0.5942
Epoch 3/5
1719/1719 [==============================] - 71s 42ms/step - loss: 1.0876 - accuracy: 0.5592 - val_loss: 0.9455 - val_accuracy: 0.6394
Epoch 4/5
1719/1719 [==============================] - 72s 42ms/step - loss: 0.8571 - accuracy: 0.6768 - val_loss: 1.0665 - val_accuracy: 0.5476
Epoch 5/5
1719/1719 [==============================] - 72s 42ms/step - loss: 0.8276 - accuracy: 0.6919 - val_loss: 0.8509 - val_accuracy: 0.6364


Not great at all, we suffered from the vanishing/exploding gradients problem.


# Batch normalization

In [15]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28,28]),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(300, activation="elu", kernel_initializer="he_normal"),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(100,activation="elu", kernel_initializer="he_normal"),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(10, activation="softmax")
])

In [16]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_2 (Flatten)         (None, 784)               0         
                                                                 
 batch_normalization (BatchN  (None, 784)              3136      
 ormalization)                                                   
                                                                 
 dense_8 (Dense)             (None, 300)               235500    
                                                                 
 batch_normalization_1 (Batc  (None, 300)              1200      
 hNormalization)                                                 
                                                                 
 dense_9 (Dense)             (None, 100)               30100     
                                                                 
 batch_normalization_2 (Batc  (None, 100)             

BNN layers adds four parameters per input: y, B(beta), u(vector of input means), o(vetcor input standard deviations),(for example the first BN layers adds, 3,136 parameters, which 4x784) the last two parameters u and o, are the moving averages; they are not affected by backpropagation, so keras calls them "non-trainable". (if you count the total number of bn parameters, 3,136+1,200 + 400, and divide by 2 , you get 2,368, which is the total number of non trainable parameters in this model)

lets take at the parameters oof the first bn layer. two are trainable (by backpropagation), and two are not:

In [18]:
#lets take at the parameters oof the first BN layer. two are trainable (by backpropagation), and two are not:
[(var.name, var.trainable) for var in model.layers[1].variables]

[('batch_normalization/gamma:0', True),
 ('batch_normalization/beta:0', True),
 ('batch_normalization/moving_mean:0', False),
 ('batch_normalization/moving_variance:0', False)]